### PR TITLE
Fix lifetime issues in test using detail::scope_exit()

### PR DESCRIPTION
**Pull request by jimmy-park**
Fix issues reported in #1534.

### CODE REVIEW

**Comment by yhirose** [comment]
@jimmy-park thank you for the changes. Could you explain why `if (thread.joinable())` is needed?

**Comment by jimmy-park** [comment]
@yhirose I added `std::thread::joinable()` for safety reasons. If some test cases need to join the thread before destruction of `detail::scope_exit()`, this will prevent double joining. I think this is a good habit when using `std::thread`.

**Comment by yhirose** [comment]
@jimmy-park I think we don't need the condition since `std::thread` objects in `test/test.cc` are managed in the default threads and are not referenced in others threads.
 
https://stackoverflow.com/questions/60339713/is-joinable-then-join-thread-safe-in-stdthread

Please let me know if my understanding isn't correct. :)

**Comment by jimmy-park** [comment]
@yhirose How about this example?
https://github.com/yhirose/cpp-httplib/blob/76230db97ff8945438ec4bb5eb7fd6630d0311a7/test/test.cc#L3942-L3965

When we apply `detail::scope_exit`, `svr.stop()` and `thread.join()` (line 3959-3960) should remain for testing closed server. So I need to check inside `detail::scope_exit` because `std::thread` is already joined or maybe somebody writes `thread.join()` in future.

Actually I handled as follows though the second joinable check is redundant at this time.
https://github.com/jimmy-park/cpp-httplib/blob/6db6723115b44c750b3daae913d3f367ee51bb31/test/test.cc#L3960-L3988

**Comment by yhirose** [comment]
@jimmy-park in this particular case, you can leave the test as it is without using `detail::scope_exit`. This is just a unit test. So we don't have to make everything 'perfect', and should be reasonable.

**Comment by jimmy-park** [comment]
@yhirose We still need `detail::scope_exit` even though it is a unit test. If `ASSERT_TRUE(res)` (line 3956) fails, the program crashes and rest tests won't run. It is even worse if we are using `std::async`, the program does not finish on failure. This happened to me when I tested on Windows.

**Comment by jimmy-park** [comment]
@yhirose Okay then I'll remove checking joinable threads.

**Comment by yhirose** [comment]
@jimmy-park thanks for your help!

File: test/test.cc [hunk 1]
```diff
@@ -2855,13 +2869,13 @@ TEST_F(ServerTest, GetStreamedEndless) {
...
       auto res = cli_.Get("/streamed-cancel",
                           [&](const char *, uint64_t) { return true; });
       ASSERT_TRUE(!res);
       EXPECT_TRUE(res.error() == Error::Canceled ||
                   res.error() == Error::Read || res.error() == Error::Write);
-    }));
+    });
   }
 
   std::this_thread::sleep_for(std::chrono::seconds(2));
```
Comment (jimmy-park) [inline test/test.cc:2881]: @yhirose For some reason, `ServerTest.ClientStop` failed and ran indefinitely (please see [results](https://github.com/yhirose/cpp-httplib/actions/runs/4539121445/jobs/7998678443)). I'm not sure what could be causing this problem. My change to `ServerTest.ClientStop` was trivial.

**Comment by yhirose** [comment]
@jimmy-park could you send another pull request?

**Pull request merged**